src/libs/portfolio/portfolioView.ts

```diff
@@ -112,10 +120,62 @@ export function calculateAccountPortfolio(
     }
   })
 
+  // For the selected account's pending state, create a SimulationNonces mapping,
+  // which associates each network with its corresponding pending simulation beforeNonce.
+  // This nonce information is crucial for determining the PendingToBeSigned or PendingToBeConfirmed Dashboard badges.
+  // For more details, see: calculatePendingAmounts.
+  const simulationNonces = Object.keys(state.pending[selectedAccount]).reduce((acc, networkId) => {
+    const beforeNonce = state.pending[selectedAccount!][networkId]?.result?.beforeNonce
+    if (beforeNonce !== undefined) {
+      acc[networkId] = beforeNonce
+    }
+
+    return acc
+  }, {} as SimulationNonces)
+
+  // We need the latest and pending token amounts for the selected account, especially for calculating the Pending badges.
+  // You might wonder why we don't retrieve this data directly from the PortfolioController. Here's the reasoning:
+  //
+  // 1. We could attach the latest amount to the controller's pending state.
+  //    However, this would mix the latest and pending data within the controller's logic, which we want to avoid.
+  //
+  // 2. Alternatively, we could fetch the latest and pending token amounts at the component level as needed.
+  //    While this seems simpler, there's a catch:
+  //    The PortfolioView is recalculated whenever certain properties change.
+  //    If we don't retrieve the latest and pending amounts within the same React update cycle,
+  //    they might become out of sync with the PortfolioView state.
+  //    Therefore, the safest and cleanest approach is to calculate these amounts during the same cycle as the PortfolioView.
+  //
+  // For more details, see: calculatePendingAmounts.
+  const tokenAmounts = Object.keys(state.latest[selectedAccount]).reduce((acc, networkId) => {
+    const latestTokens = state.latest[selectedAccount!][networkId]?.result?.tokens
+
+    if (!latestTokens) return acc
+
+    const mergedTokens = latestTokens.map((latestToken) => {
+      const pendingToken = state.pending[selectedAccount!][networkId]?.result?.tokens.find(
+        (pending) => {
+          return pending.address === latestToken.address
+        }
+      )
+
+      return {
+        latestAmount: latestToken.amount || 0n,
+        pendingAmount: pendingToken?.amount || 0n,
+        address: latestToken.address,
+        networkId
+      }
+    })
+
+    return [...acc, ...mergedTokens]
+  }, [] as TokenAmount[])
+
```

Comment:
Me neither, but it's the most compromiseable solution given the current controllers and portfolioView structure.

However, after reviewing portfolioView and getTokenDetails, as well as the latest changes we made to always fetch the pending and latest portfolio states, there is a better way to handle all the layers. This should be part of a separate PR, as it will introduce a significant number of changes.

How can we do this?

Keeping in mind that we always prioritize the pending state over the latest, we can simplify what portfolioView is computing and consistently rely on the pending state.

Additionally, since we always use the pending state, the return values from the getTokenDetails helper are somewhat redundant. Currently, we return balance, balanceFormatted, pendingBalance, and pendingBalanceFormatted, but in the current setup (if I'm not mistaken), these always resolve to the pending state.

With this in mind, we can refactor portfolioView to simply return the pending state and include the pending calculations from getTokenDetails. By doing this, we will no longer need to manually call getTokenDetails in all the components, and we can eliminate the tokenAmounts derived state.

Reply:
We discussed it offline and agreed on the above comment and proposal.